vpr/src/pack/cluster_legalizer.h

```diff
@@ -0,0 +1,478 @@
+#pragma once
+
+#include <set>
+#include <unordered_map>
+#include "atom_netlist_fwd.h"
+#include "noc_data_types.h"
+#include "partition_region.h"
+#include "vpr_types.h"
+#include "vtr_range.h"
+#include "vtr_strong_id.h"
+#include "vtr_vector.h"
+#include "vtr_vector_map.h"
+
+class Prepacker;
+class t_pb_graph_node;
+struct t_lb_router_data;
+
+// A special ID to identify the legalization clusters. This is separate from the
+// ClusterBlockId since this legalizer is not necessarily tied to the Clustered
+// netlist, but is used as a sub-routine to it.
+struct legalization_cluster_id_tag;
+typedef vtr::StrongId<legalization_cluster_id_tag, size_t> LegalizationClusterId;
+
+/// @brief The different legalization strategies the cluster legalizer can perform.
+///
+/// Allows the user of the API to select how thorough the legalizer should be
+/// when adding molecules into clusters.
+enum class ClusterLegalizationStrategy {
+    FULL,                   // Run the full legalizer (including intra-lb routing)
+    SKIP_INTRA_LB_ROUTE     // Do all legality checks except intra-lb routing
+};
+
+/// @brief The status of the cluster legalization.
+enum class e_block_pack_status {
+    BLK_PASSED,                 // Passed legalization.
+    BLK_FAILED_FEASIBLE,        // Failed due to block not feasibly being able to go in the cluster.
+    BLK_FAILED_ROUTE,           // Failed due to intra-lb routing failure.
+    BLK_FAILED_FLOORPLANNING,   // Failed due to not being compatible with the cluster's current PartitionRegion.
+    BLK_FAILED_NOC_GROUP,       // Failed due to not being compatible with the cluster's NoC group.
+    BLK_STATUS_UNDEFINED        // Undefined status. Something went wrong.
+};
+
+/*
+ * @brief A struct containing information about the cluster.
+ *
+ * This contains necessary information for legalizing a cluster.
+ */
+struct LegalizationCluster {
+    /// @brief A list of the molecules in the cluster. By design, a cluster will
+    ///        only contain molecules which have been previously legalized into
+    ///        the cluster using a legalization strategy.
+    std::set<t_pack_molecule*> molecules;
```

Comment:
Originally I wanted to allow clusters to be modified (removing a molecule from a cluster); however, I found that the feature was never used (and would never be tested). This should have been an unordered_set anyways. I turned it into a vector, since it is only used in the cluster_legalizer to destroy the cluster, which requires that we just iterate over the container.